File: personal/investments/scripts/investments/render.py
"""Render the masked datastore and analytics into one self-contained offline page.

"The Ledger": a classical broadsheet dashboard adapted from the imported Claude
Design "classical" system. A single filter-driven page: a global filter bar
(accounts grouped by kind, plus a date range) drives every headline figure and
chart. Fonts are self-hosted, all figures and charts are recomputed client-side
by ledger.js from one embedded dataset, no network at view time.

The statements record contributions, income, cash balances and cost basis, not
live market value; every figure is stated at cost or as cash on hand and never
implies a market quote.
"""

import re
from html import escape as _std_escape
from json import dumps as _json_dumps
from pathlib import Path

CSS_HREF = "../../_assets/personal.css"

_KIND_ORDER = [
    "TFSA",
    "ManagedTFSA",
    "FHSA",
    "RRSP",
    "RESP",
    "NonRegistered",
    "DirectIndexing",
    "PE",
    "Crypto",
    "Other",
    "Chequing",
    "Savings",
    "USD",
    "CreditCard",
]

STANDING_NOTE = (
    "Compiled from monthly statements, which record what was paid, not what "
    "holdings are worth. Figures are stated at cost; the brokerage app shows "
    "market value, so its totals will differ, usually upward."
)

_LEDGER_JS = (Path(__file__).parent / "ledger.js").read_text(encoding="utf-8")


def _html_escape(text: object) -> str:
    """Escape any user-derived value for safe inclusion in HTML text or attributes."""
    return _std_escape(str(text), quote=True)


def _slug(text: str) -> str:
    """Lowercase hyphenated slug for an account id fragment."""
    return re.sub(r"[^a-z0-9]+", "-", text.lower()).strip("-")


def _account_id_label(a: dict) -> str:
    """The requested kind-account-name plus short id machine label."""
    return f"{_slug(a['kind'])}-{_slug(a['name'])}·{a['short_id']}"


def _kind_sort(kind: str) -> int:
    return _KIND_ORDER.index(kind) if kind in _KIND_ORDER else len(_KIND_ORDER)


def _filter_bar(accounts: list[dict]) -> str:
    """Global filter bar: account chips grouped by kind, and a date-range control."""
    ordered = sorted(accounts, key=lambda a: (_kind_sort(a["kind"]), a["name"]))
    groups: dict[str, list[dict]] = {}
    for a in ordered:
        groups.setdefault(a["kind"], []).append(a)
    chip_groups = []
    for kind, accts in groups.items():
        chips = "".join(
            f'<button type="button" class="chip acct" data-acct="{_html_escape(a["masked_id"])}" '
            f'data-kind="{_html_escape(kind)}" title="{_html_escape(_account_id_label(a))}">'
            f'{_html_escape(a["name"])} <span class="chip-id">{_html_escape(a["short_id"])}'
            "</span></button>"
            for a in accts
        )
        chip_groups.append(
            '<div class="kind-group">'
            f'<button type="button" class="chip kind-toggle" data-kind="{_html_escape(kind)}">'
            f'<span class="badge">{_html_escape(kind)}</span></button>{chips}</div>'
        )
    return (
        '<div class="filterbar" id="filterbar"><div class="fb-accounts">'
        '<button type="button" class="chip on" data-all>All accounts</button>'
        f"{''.join(chip_groups)}</div>"
        '<div class="fb-side"><div class="seg fb-dates" role="group" aria-label="Date range">'
        '<button type="button" class="seg-btn" data-range="ytd">YTD</button>'
        '<button type="button" class="seg-btn" data-range="1y">1Y</button>'
        '<button type="button" class="seg-btn" data-range="3y">3Y</button>'
        '<button type="button" class="seg-btn on" data-range="all">All</button></div>'
        '<div class="fb-scope" id="fb-scope"></div></div></div>'
    )


def _masthead() -> str:
    """Newspaper masthead with the standing editorial note about the data source."""
    return (
        '<header class="masthead">'
        '<div class="masthead-kicker">Personal Finance · Household Edition</div>'
        '<h1 class="masthead-title">The Ledger</h1></header>'
        '<div class="dateline"><span>Private Records</span>'
        '<span class="reviewed" id="asof">Portfolio report</span>'
        "<span>Stated at cost</span></div>"
        f'<p class="standing-note">{_html_escape(STANDING_NOTE)}</p>'
    )


def _chart_section(title: str, note: str, chart_id: str, extra: str = "") -> str:
    """A section with a heading, note, and an interactive chart mount point."""
    return (
        f'<section class="section"><div class="section-head"><div>'
        f'<h2 class="section-title">{_html_escape(title)}</h2>'
        f'<p class="section-note">{_html_escape(note)}</p></div>{extra}</div>'
        f'<div class="chartbox" id="{chart_id}"><div class="ex-tip" id="{chart_id}-tip"></div>'
        f'<div class="chart-ymax" id="{chart_id}-ymax"></div>'
        f'<div class="chart-ymin" id="{chart_id}-ymin"></div></div>'
        f'<div class="chart-legend" id="{chart_id}-legend"></div></section>'
    )


def _footnote() -> str:
    """The classical about-these-figures footnote at the page foot."""
    return (
        '<section class="section footnote" id="about"><div class="section-head"><div>'
        '<h2 class="section-title">About these figures</h2></div></div>'
        "<p>These figures are compiled from monthly account statements. The statements record "
        "cash movements (deposits, withdrawals, transfers), contributions, dividends and "
        "interest received, and each buy or sell with its price at the time of trade. They do "
        "not carry current market prices, so a holding's present market value and any "
        "unrealised gain cannot be shown.</p>"
        "<p><strong>Invested at cost</strong> is the adjusted cost base of positions still held: "
        "what was paid for them, reduced proportionally as they were sold. "
        "<strong>Contributions</strong> are deposits coded as contributions, gross of "
        "recontributions. <strong>Cash on hand</strong> is uninvested cash. "
        "<strong>Growth beyond contributions</strong> is the cost base above contributions, i.e. "
        "capital that arrived as transfers in or reinvested income.</p></section>"
    )


def render_index(store: dict, analytics: dict) -> str:
    """The single filter-driven Ledger page (all figures computed by ledger.js)."""
    body = (
        _masthead()
        + _filter_bar(store["accounts"])
        + '<div class="hero-row" id="headline"></div>'
        + '<div class="waterfall" id="waterfall"></div>'
        + '<div class="hr mt-rule"></div>'
        + _chart_section(
            "Capital deployed over time",
            "Cumulative cost base of holdings plus cash. The thin line is cumulative "
            "contributions; the gap above it is transfers in and reinvested income.",
            "cap",
        )
        + '<div class="hr mt-rule"></div>'
        + _chart_section("Income received", "Dividends and interest by period.", "inc")
        + '<div class="hr mt-rule"></div>'
        + _chart_section(
            "Cash flow", "Money in above the line, money out below; net marked per period.", "cf"
        )
        + '<div class="hr mt-rule"></div>'
        + '<section class="section"><div class="section-head"><div>'
        '<h2 class="section-title">Contributions &amp; room</h2>'
        '<p class="section-note">Contributed against annual room, per registered type.</p>'
        '</div></div><div id="room"></div></section>'
        + '<div class="hr mt-rule"></div>'
        + '<section class="section"><div class="section-head"><div>'
        '<h2 class="section-title">Accounts</h2>'
        '<p class="section-note">Grouped by type; figures within the current filter.</p>'
        '</div></div><div class="table-wrap" id="acct-table"></div></section>'
        + '<div class="hr mt-rule"></div>'
        + '<section class="section"><div class="section-head"><div>'
        '<h2 class="section-title">Holdings at cost</h2>'
        '<p class="section-note">Adjusted cost base, not market value. Direct Indexing '
        "is collapsed to one row.</p></div></div>"
        '<div class="table-wrap" id="hold-table"></div></section>'
        + '<div class="hr mt-rule"></div>'
        + _footnote()
    )
    payload = {"ledger": analytics["ledger"]}
    foot = (
        f'<script type="application/json" id="ledger-data">{_json_dumps(payload)}</script>'
        f"<script>{_LEDGER_JS}</script>"
    )
    return _page("The Ledger", body, foot)


def _page(title: str, body: str, foot: str = "") -> str:
    """Wrap a body in the full offline document shell."""
    return (
        '<!DOCTYPE html>\n<html lang="en">\n<head>\n<meta charset="utf-8">\n'
        '<meta name="viewport" content="width=device-width, initial-scale=1">\n'
        f"<title>{_html_escape(title)} — The Ledger</title>\n"
        f'<link rel="stylesheet" href="{CSS_HREF}">\n</head>\n<body>\n'
        f'<main class="page">{body}</main>{foot}\n</body>\n</html>\n'
    )


def render_pages(store: dict, analytics: dict) -> dict[str, str]:
    """Render the single Ledger page keyed by filename."""
    return {"index.html": render_index(store, analytics)}

File: personal/investments/scripts/src/render.ts
// Render the datastore and analytics into one self-contained offline page.
import { readFileSync } from "node:fs";
import type { Analytics } from "./analytics";
import { monthLabel } from "./client/format";
import type { Datastore } from "./datastore";

const CSS_HREF = "../../_assets/personal.css";
const FLATPICKR_CSS_PATH = new URL(
  "../node_modules/flatpickr/dist/flatpickr.min.css",
  import.meta.url,
);

const STANDING_NOTE =
  "Compiled from monthly statements, which record what was paid, not what holdings are " +
  "worth. Figures are stated at cost; the brokerage app shows market value, so its totals " +
  "will differ, usually upward.";

async function bundleClient(): Promise<string> {
  const built = await Bun.build({
    entrypoints: [new URL("./client/main.ts", import.meta.url).pathname],
    minify: true,
    target: "browser",
  });
  const out = built.outputs[0];
  if (!out) throw new Error("client bundle produced no output");
  return await out.text();
}

function htmlEscape(text: unknown): string {
  return String(text)
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&#x27;");
}

// Static skeleton for the progressive scope filter. filter.ts populates
// #account-groups from the embedded ledger data at runtime and wires all
// interactions; this only emits the summary button and the empty popover
// shell so the page has something to attach to before the client script
// runs. Selectors are namespaced "fbx-" so they never collide with the rest
// of the shared stylesheet.
function filterShell(): string {
  return (
    '<div class="fbx-bar">' +
    '<button type="button" class="fbx-summary" id="scope-summary" aria-haspopup="true" ' +
    'aria-expanded="false" aria-controls="scope-popover">' +
    '<span id="scope-summary-text">All accounts · All time</span>' +
    '<span class="fbx-caret" aria-hidden="true">&#9662;</span></button>' +
    '<div class="fbx-popover" id="scope-popover" hidden>' +
    '<div class="fbx-popover-col"><div class="fbx-popover-label">Accounts</div>' +
    '<div id="account-groups"></div></div>' +
    '<div class="fbx-popover-col"><div class="fbx-popover-label">Time window</div>' +
    '<div class="fbx-presets" role="group" aria-label="Time presets">' +
    '<button type="button" class="fbx-preset on" data-preset="all">All</button>' +
    '<button type="button" class="fbx-preset" data-preset="ytd">YTD</button>' +
    '<button type="button" class="fbx-preset" data-preset="1y">1Y</button>' +
    '<button type="button" class="fbx-preset" data-preset="3y">3Y</button></div>' +
    '<input type="text" class="fbx-range-input" id="range-picker" ' +
    'placeholder="Custom range" readonly></div></div></div>'
  );
}

// "Jul 2023 – Jun 2026" span of the underlying data, from the first to the
// last month in the ledger. Rendered statically at build time since the
// month list never changes without a rebuild.
function dataRangeLabel(months: string[]): string {
  const first = months[0];
  const last = months[months.length - 1];
  if (!first || !last) return "";
  return `Data: ${monthLabel(first)} – ${monthLabel(last)}`;
}

function masthead(months: string[]): string {
  return (
    '<header class="masthead">' +
    '<div class="masthead-kicker">Personal Finance · Household Edition</div>' +
    '<h1 class="masthead-title">The Ledger</h1></header>' +
    '<div class="dateline"><span>Private Records</span>' +
    '<span class="reviewed" id="asof">Portfolio report</span>' +
    `<span id="data-range">${htmlEscape(dataRangeLabel(months))}</span>` +
    "<span>Stated at cost</span></div>" +
    `<p class="standing-note">${htmlEscape(STANDING_NOTE)}</p>`
  );
}

function canvasBox(chartId: string): string {
  return `<div class="chartbox-canvas"><canvas id="${chartId}"></canvas></div>`;
}

function sectionHead(title: string, note: string, titleId?: string): string {
  const idAttr = titleId ? ` id="${titleId}"` : "";
  return (
    '<div class="section-head"><div>' +
    `<h2 class="section-title"${idAttr}>${htmlEscape(title)}</h2>` +
    `<p class="section-note">${htmlEscape(note)}</p></div></div>`
  );
}

function subhead(title: string): string {
  return `<h3 class="pillar-subhead">${htmlEscape(title)}</h3>`;
}

function contribSection(): string {
  return (
    '<section class="section" id="section-contrib">' +
    sectionHead(
      "Contributions & Room",
      "Money you put in and registered room used across the selected accounts and time " +
        "window.",
    ) +
    '<div class="hero-row" id="headline"></div>' +
    '<div id="room"></div>' +
    subhead("External money in / out") +
    '<p class="section-note">External deposits into and withdrawals out of the selected ' +
    "accounts, not trading activity — transfers between your own accounts are excluded. " +
    "Click a bar to see the underlying transactions.</p>" +
    canvasBox("chart-cashflow") +
    '<details class="pillar-detail" id="cashflow-drill">' +
    "<summary>Transactions for the selected period</summary>" +
    '<div id="cashflow-drill-body"></div></details>' +
    "</section>"
  );
}

function growthSection(): string {
  return (
    '<section class="section" id="section-growth">' +
    sectionHead(
      "Growth",
      "Net deposits, portfolio at cost, and income received across the selected accounts " +
        "and time window.",
    ) +
    '<div class="hero-row" id="growth-summary"></div>' +
    subhead("Portfolio at cost vs net deposits") +
    canvasBox("chart-trend") +
    subhead("By account, at cost") +
    canvasBox("chart-growth") +
    "</section>"
  );
}

function taxSection(): string {
  return (
    '<section class="section" id="section-tax">' +
    sectionHead(
      "Tax",
      "Interest, eligible dividends, foreign income, and realized gains for the selected " +
        "accounts, for the selected window's tax year.",
      "tax-section-title",
    ) +
    '<div class="hero-row" id="tax-cards"></div>' +
    '<div class="pillar-tax-rate">' +
    '<label for="tax-rate">Marginal rate</label>' +
    '<input type="number" id="tax-rate" step="0.01" min="0" max="1" value="0.48">' +
    '<span>Estimated tax added: <strong id="tax-estimate"></strong></span></div>' +
    '<p class="pillar-disclaimer"><strong>Rough estimate — not for filing.</strong></p>' +
    subhead("Income received") +
    canvasBox("chart-income") +
    "</section>"
  );
}

function projectionSection(): string {
  return (
    '<section class="section" id="section-projection">' +
    sectionHead(
      "Thirty Year Projection",
      "A forward projection of contributions, government grants, room remaining, and " +
        "projected value across the selected accounts, driven by the rate inputs below. " +
        "Registered accounts follow their CRA limits; the corporate account follows a " +
        "flat plan, since a corporation has no contribution room.",
      "projection-section-title",
    ) +
    '<div class="pillar-tax-rate">' +
    '<label for="proj-return">Return</label>' +
    '<input type="number" id="proj-return" step="0.1" min="0" max="20" value="8">' +
    '<label for="proj-index">Indexation</label>' +
    '<input type="number" id="proj-index" step="0.1" min="0" max="10" value="2">' +
    "</div>" +
    '<div id="proj-warning"></div>' +
    '<div class="hero-row" id="proj-summary"></div>' +
    subhead("Year by year") +
    '<div class="table-wrap" id="proj-table"></div>' +
    subhead("Contributions, grants, and growth") +
    canvasBox("chart-projection") +
    subhead("Projected value by account") +
    '<p class="section-note">Contribution room is assessed per person and per group, not per ' +
    "account, so a group's room is split across its accounts by their recent share of that " +
    "group's contributions. A line keeps its colour whatever else is selected. " +
    "The value axis is logarithmic, so steady growth reads as a straight line and " +
    "accounts of very different sizes stay comparable \u2014 but equal vertical " +
    "distances mean equal <em>ratios</em>, not equal dollars. A line that stops has " +
    "closed.</p>" +
    canvasBox("chart-projection-accounts") +
    '<div id="proj-caveats"></div>' +
    "</section>"
  );
}

function detailSection(): string {
  return (
    '<section class="section" id="section-detail">' +
    sectionHead("Detail", "Accounts and holdings within the current scope.") +
    '<details class="pillar-detail"><summary>Show tables</summary>' +
    subhead("Accounts") +
    '<p class="section-note">All figures at cost; the holdings table below breaks holdings out ' +
    "by security.</p>" +
    '<div class="table-wrap" id="acct-table"></div>' +
    subhead("Holdings at cost") +
    '<p class="section-note">Adjusted cost base, not market value. Direct Indexing is ' +
    "collapsed to one row.</p>" +
    '<div class="table-wrap" id="hold-table"></div>' +
    "</details></section>"
  );
}

function footnote(): string {
  return (
    '<section class="section footnote" id="about"><div class="section-head"><div>' +
    '<h2 class="section-title">About these figures</h2></div></div>' +
    "<p>These figures are compiled from monthly account statements. The statements record " +
    "cash movements (deposits, withdrawals, transfers), contributions, dividends and " +
    "interest received, and each buy or sell with its price at the time of trade. They do " +
    "not carry current market prices, so a holding's present market value and any " +
    "unrealised gain cannot be shown.</p>" +
    "<p><strong>Portfolio at cost</strong> is the adjusted cost base of positions still held " +
    "plus cash on hand: what was paid for them, reduced proportionally as they were sold, " +
    "plus uninvested cash. <strong>Registered contributions</strong> are deposits coded as " +
    "contributions, gross of recontributions, used only for CRA contribution room. " +
    "<strong>Net deposits</strong> is the true measure of money put in: it adds transfers in " +
    "and subtracts transfers out on top of registered contributions, so it also counts money " +
    "that arrived as a transfer rather than a coded contribution. It is the closest match to " +
    "the brokerage app's own net-deposits figure, though the statements cannot tell an " +
    "internal transfer between your own accounts from an external withdrawal, so the two need " +
    "not tie to the cent. <strong>Gain beyond deposits</strong> is portfolio at cost minus net " +
    "deposits: cost basis above what was put in, i.e. reinvested income and securities " +
    "transferred in at cost. It is a cost-basis figure, not a market-value gain.</p></section>"
  );
}

function rule(): string {
  return '<div class="hr mt-rule"></div>';
}

function page(title: string, body: string, foot: string): string {
  const flatpickrCss = readFileSync(FLATPICKR_CSS_PATH, "utf-8");
  return (
    '<!DOCTYPE html>\n<html lang="en">\n<head>\n<meta charset="utf-8">\n' +
    '<meta name="viewport" content="width=device-width, initial-scale=1">\n' +
    `<title>${htmlEscape(title)} — The Ledger</title>\n` +
    `<link rel="stylesheet" href="${CSS_HREF}">\n` +
    `<style>${flatpickrCss}</style>\n</head>\n<body>\n` +
    `<main class="page">${body}</main>${foot}\n</body>\n</html>\n`
  );
}

export async function renderIndex(_store: Datastore, analytics: Analytics): Promise<string> {
  const body =
    masthead(analytics.ledger.months) +
    filterShell() +
    contribSection() +
    rule() +
    growthSection() +
    rule() +
    taxSection() +
    rule() +
    projectionSection() +
    rule() +
    detailSection() +
    rule() +
    footnote();
  const payload = JSON.stringify({ ledger: analytics.ledger }).replace(/</g, "\\u003c");
  const clientJs = await bundleClient();
  const foot =
    `<script type="application/json" id="ledger-data">${payload}</script>` +
    `<script>${clientJs}</script>`;
  return page("The Ledger", body, foot);
}

export async function renderPages(
  store: Datastore,
  analytics: Analytics,
): Promise<Record<string, string>> {
  return { "index.html": await renderIndex(store, analytics) };
}
